### ISSUE TITLE
[Proposal] Manual Database Connection for Tenants

### ISSUE THREAD

**Issue by johannesschobel** [tenancy/tenancy#12]
Dear @luceos ,

I am proposing a valuable request for the upcoming `tenancy` package - at least i think, this may be valuable, depending on the actual application scenario.

- In the application i will need to develop, we need to follow a _one tenant per database_ approach (e.g., you will typically have `n+1` databases (`n` tenant databases and `1` system database).

- We will need to "distribute" the tenant databases to different servers (e.g., `host = localhost` is not suitable). This will also be raised by issue #5 

- We will need to "manually" assign the database connection properties (e.g., database name, user & password) to a specific tenant because of several reasons (e.g., the hoster automatically creates a database name, we are not able to change!). In the given application scenario it is not suitable to use the `Generator` to create databases, db-users and corresponding passwords.

-------

I think, it would be awesome to add a few different fields on the `Tenant` model. For example, `db_connection`, `db_host`, `db_name`, `db_user`, `db_password`. If respective fields are set to `null` (for example), the default `Generator` may be applied.

I initially described this issue in the `hyn/multi-tenancy` package as well - the related issue is here: https://github.com/hyn/multi-tenant/issues/557

What do you think on this one? Do you think, this proposal could make it into this package?

All the best and thank you very much for your time

**Comment by hafael** [comment]
I also have the same need. I installed and configured the package just for this purpose.
I'm following this topic.

**Comment by luceos** [comment]
First off this package is not ready, I even recently removed some crucial code which handled eloquent database connection.

Related to the question, compared to hyn, tenancy will not hold the Tenant object anymore. There won't be any migrations for the Tenant.  You can simply create your own implementation of a Tenant, apply the trait and register it. You can then have your own database resolving logic read the mentioned columns from the Tenant eloquent model (you use) to set up a connection.

tldr your proposal should be no issue at all.

For information about the assignment of a Tenant check the (wip) docs: https://laravel-tenancy.com/docs/tenancy/1.x/what-is-a-tenant

**Comment by johannesschobel** [comment]
Dear @luceos ,
Thank you for your detailed explanation on this topic.. glad to hear that you have this on your mind as well :+1: 

All the best with your awesome work!

**Comment by rubensrocha** [comment]
I also support this idea, mainly so that it is possible to use this package in shared hosting (I know the focus right now are vps/dedicated). If there is a way to tell the connection data for each tenant (you can use each tentant's folder in storage/app/tenancy/tenants) it would make it easier to use this package in this scenario.

**Comment by luceos** [comment]
Now that t/t (tenancy/tenancy) is reaching a mature state I was going over this request again.

I would implement this like this:

- create a tenant implementing the required contracts/interfaces that has the columns needed to set up the database connection.
- create a custom db driver (from scratch or by forking one of the existing ones).
- use the custom db driver instead of the default ones.

This sounds simple and straightforward. I'm not going to test this for now, feel free to report back with findings once you gave it a shot.
